tfhe/src/integer/server_key/radix_parallel/mod.rs

```diff
@@ -88,36 +88,94 @@ impl ServerKey {
         carry
     }
 
-    pub fn partial_propagate_parallelized<T>(&self, ctxt: &mut T, start_index: usize)
+    /// Propagates carries starting from start_index.
+    ///
+    /// Does nothing if start_index >= ctxt.len() or ctxt is empty
+    pub fn partial_propagate_parallelized<T>(&self, ctxt: &mut T, mut start_index: usize)
     where
         T: IntegerRadixCiphertext,
     {
-        if self.is_eligible_for_parallel_single_carry_propagation(ctxt) {
-            let num_blocks = ctxt.blocks().len();
+        if start_index >= ctxt.blocks().len() || ctxt.blocks().is_empty() {
+            return;
+        }
 
-            let (mut message_blocks, carry_blocks) = rayon::join(
+        // Extract message blocks and carry blocks from the
+        // input block slice.
+        // Carries Vec has one less block than message Vec
+        let extract_message_and_carry_blocks = |blocks: &[crate::shortint::Ciphertext]| {
+            let num_blocks = blocks.len();
+
+            rayon::join(
                 || {
-                    ctxt.blocks()[start_index..]
+                    blocks
                         .par_iter()
                         .map(|block| self.key.message_extract(block))
                         .collect::<Vec<_>>()
                 },
                 || {
                     let mut carry_blocks = Vec::with_capacity(num_blocks);
                     // No need to compute the carry of the last block, we would just throw it away
-                    ctxt.blocks()[start_index..num_blocks - 1]
+                    blocks[..num_blocks - 1]
                         .par_iter()
                         .map(|block| self.key.carry_extract(block))
                         .collect_into_vec(&mut carry_blocks);
-                    carry_blocks.insert(0, self.key.create_trivial(0));
                     carry_blocks
                 },
-            );
+            )
+        };
+
+        if self.is_eligible_for_parallel_single_carry_propagation(ctxt) {
+            let highest_degree = ctxt.blocks()[start_index..]
+                .iter()
+                .max_by(|block_a, block_b| block_a.degree.get().cmp(&block_b.degree.get()))
+                .map(|block| block.degree.get())
+                .unwrap(); // We checked for emptiness earlier
+            if highest_degree <= (self.key.message_modulus.0 - 1) * 2 {
```

Comment:
self.key.message_modulus.0.next_power_of_2().next_power_of_2(), will yield 4 if for 2_2, if you meant 8, it would be to high

The constrain here is that, at most each block, can only have 1 bit of carry taken and its message part shall allow adding `1` without creating a new carry, which for 2_2 is 6, 3 * 2

For example we have as input a radix:


```
MSB -> LSB
[00|00] [01|11] [01|00] <=> 0, 7, 4
```
the carry of the first block, when added to the second one will change the block value to
`[10|00]` but the parallel algorithm won't catch that and will still say the carry out of block 2nd
is `01`

It might be possible to modify the algorithm to allow that but its not the case right now

Reply:
ah yes it's not only the carry bit, it's being able to add 1 full message as well